src/app/services/WebGL/web-gl.service.ts

```diff
@@ -0,0 +1,188 @@
+})
+export abstract class WebGLService {
+
+
+  private canvasSize: Vector2 | undefined;
+  private canvasClearColor: Color | undefined;
+
+  private _renderingContext: CanvasRenderingContext2D | ImageBitmapRenderingContext | WebGLRenderingContext | WebGL2RenderingContext | null | undefined;
+
+  protected get gl(): WebGLRenderingContext {
+    return this._renderingContext as WebGLRenderingContext;
+  }
+
+  protected vertexShader: WebGLShader | undefined;
+  protected fragmentShader: WebGLShader | undefined;
+
+  private _program: WebGLProgram | undefined;
+  protected get program(): WebGLProgram {
+    return <WebGLProgram>this._program;
+  }
+
+  private _buffer: WebGLBuffer | undefined;
+  protected get buffer(): WebGLBuffer {
+    return <WebGLBuffer>this._buffer;
+  }
+
+  protected constructor() {
+    //empty
+  }
+
+  /**
+   * Initializes the WebGL Rendering Context and Canvas
+   * @param canvas the HTML Canvas Object
+   * @param canvasSize the dimensions of the Canvas
+   * @param clearColor the color the canvas is reset to when cleared
+   */
+  public initializeRenderingContext(canvas: HTMLCanvasElement, canvasSize: Vector2, clearColor: Color): boolean {
+    this._renderingContext = canvas.getContext('webgl');
+    if(!this.gl) {
+      console.log("Unable to initialize WebGL. Your browser may not support it.");
+      return false;
+    }
+
+    this.canvasSize = canvasSize;
+    this.canvasClearColor = clearColor;
+
+    this.setCanvasSize(this.canvasSize.x, this.canvasSize.y);
+    this.setCanvasClearColor(this.canvasClearColor);
+    this.clearCanvas();
+
+    return true;
+  }
+
+  protected setCanvasSize(x: number, y: number): void {
+    this.gl.canvas.width = x;
+    this.gl.canvas.height = y;
+  }
+
+  protected setCanvasClearColor(color: Color): void {
+    this.gl.clearColor(color.r, color.g, color.b, color.a);
+  }
+
+  public clearCanvas(): void {
+    this.gl.clear(this.gl.COLOR_BUFFER_BIT | this.gl.DEPTH_BUFFER_BIT);
+  }
+
+  /**
+   * Initializes all of the service's WebGL components
+   * @private
+   * @return boolean specifies whether initialization was successful
+   */
+  protected initialize(vertexSrc: string, fragmentSrc: string): boolean {
+    if(this.isInitialized()){
+      return true;
+    }
+
+    //Initialize Shaders
+    this.initializeShader(this.gl.VERTEX_SHADER, vertexSrc);
+    this.initializeShader(this.gl.FRAGMENT_SHADER, fragmentSrc);
+
+    if(!this.vertexShader || !this.fragmentShader){
+      return false;
+    }
+
+    //Initialize Program
+    this.initializeProgram(this.vertexShader, this.fragmentShader);
+
+    if(!this.program){
+      return false;
+    }
+
+    this.initializeBuffer();
+
+    return this.buffer != null;
+  }
+
+  private isInitialized(): boolean {
+    return this.gl != null
+      && this.vertexShader != null
+      && this.fragmentShader != null
+      && this.program != null
+      && this.buffer != null;
+  }
+
+  protected updateContext(){
```

Comment:
Github is wanting a return type here. Is it right?

Reply:
Think it is implicitly turning to void, but should make it void explicit

Reply:
Yeah, the linter caught that. Should be fixed now.

Reply:
blessssss